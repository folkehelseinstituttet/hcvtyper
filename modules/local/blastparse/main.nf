process BLASTPARSE {
    tag "$meta.id"
    label 'process_single'

    // Environment with R tidyverse and seqinr packages from the conda-forge channel. Created using seqera containers.
    // URLs:
    // Docker image: https://wave.seqera.io/view/builds/bd-5358395134867368_1?_gl=1*1vaclhd*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    // Singularity image: https://wave.seqera.io/view/builds/bd-0225dab2b8112adf_1?_gl=1*111m8r6*_gcl_au*MTkxMjgxNTMwMi4xNzUzNzczOTQz
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/b5/b5d7b977f4b94903794f65bbd852248eef87ce608c5eef34d605afabf514f397/data':
        'community.wave.seqera.io/library/r-seqinr_r-tidyverse:5358395134867368' }"

    input:
    tuple val(meta), path(blast_out), path(contigs)
    path(references)
    val(agens)

    output:
    tuple val(meta), path("*contigs.fa")    , emit: contigs    , optional: true
    tuple val(meta), path('*blast_out.csv') , emit: blast_res
    tuple val(meta), path("*major.fa")      , emit: major_fasta, optional: true
    tuple val(meta), path("*minor.fa")      , emit: minor_fasta, optional: true
    tuple val(meta), path("*blastparse.csv"), emit: csv
    tuple val(meta), path("*.png")          , emit: png
    path "versions.yml"                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def agens  = task.ext.agens ?: "${agens}"

    """
    blast_parse.R \\
        $prefix \\
        $blast_out \\
        $contigs \\
        $references \\
        $agens \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
      r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
      tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
      seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    echo "${args}"

    # Deterministic stub outputs matching declared outputs
    printf "qseqid,sseqid,pident,length,mismatch,gapopen,qstart,qend,sstart,send,evalue,bitscore\n" > ${prefix}.blast_out.csv
    printf "id,header\n" > ${prefix}.blastparse.csv
    : > ${prefix}.png

    # Optional outputs (safe to keep empty)
    : > ${prefix}.contigs.fa
    : > ${prefix}.major.fa
    : > ${prefix}.minor.fa

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
      r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
      tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
      seqinr: \$(Rscript -e "library(seqinr); cat(as.character(packageVersion('seqinr')))")
    END_VERSIONS
    """
}
