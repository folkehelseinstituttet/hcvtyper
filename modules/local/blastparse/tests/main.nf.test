// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core modules test blastparse
nextflow_process {

  name "Test Process BLASTPARSE"
  script "../main.nf"
  process "BLASTPARSE"

  tag "modules"
  tag "blastparse"

  // Real, tiny run (requires docker/conda)
  test("blastparse: blast6 + contigs + references") {
    options "--profile docker"
    when {
      process {
        """
        // Input signature:
        // tuple val(meta), path(blast_out), path(contigs)
        // path(references)
        // val(agens)

        input[0] = [
          [ id: 'toy' ],
          file("${projectDir}/modules/local/blastparse/tests/blastout.txt",       checkIfExists: true),
          file("${projectDir}/modules/local/blastparse/tests/Test.contigs.fa.gz", checkIfExists: true)
        ]
        input[1] = file("${projectDir}/modules/local/blastparse/tests/references.fa", checkIfExists: true)
        input[2] = 'HCV'
        """
      }
    }
    then {
      assertAll(
        { assert process.success },
        // Snapshot only what you care about
        { assert snapshot([
            blast_res: process.out.blast_res,
            csv      : process.out.csv,
            png      : process.out.png,
            versions : process.out.versions
          ]).match() }
      )
    }
  }

  // Stub run (fast)
  test("blastparse: stub") {
    options "-stub-run"
    when {
      process {
        """
        input[0] = [
          [ id: 'toy' ],
          file("${projectDir}/modules/local/blastparse/tests/blastout.txt",       checkIfExists: true),
          file("${projectDir}/modules/local/blastparse/tests/Test.contigs.fa.gz", checkIfExists: true)
        ]
        input[1] = file("${projectDir}/modules/local/blastparse/tests/references.fa", checkIfExists: true)
        input[2] = 'HCV'
        """
      }
    }
    then {
      assertAll(
        { assert process.success },
        { assert snapshot(process.out).match() }
      )
    }
  }
}
