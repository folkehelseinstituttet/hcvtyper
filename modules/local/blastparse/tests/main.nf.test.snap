{
    "blastparse: stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.blast_out.csv:md5,854d3f407b6d8747e8dc79862c911522"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.major.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.minor.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.blastparse.csv:md5,cd5ec569f4a4411f5406d5222d2de0ef"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,1618d74e63656df309db763cabd3556a"
                ],
                "blast_res": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.blast_out.csv:md5,854d3f407b6d8747e8dc79862c911522"
                    ]
                ],
                "contigs": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.contigs.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "csv": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.blastparse.csv:md5,cd5ec569f4a4411f5406d5222d2de0ef"
                    ]
                ],
                "major_fasta": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.major.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "minor_fasta": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.minor.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "png": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1618d74e63656df309db763cabd3556a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-06T11:24:56.058643387"
    },
    "blastparse: blast6 + contigs + references": {
        "content": [
            {
                "blast_res": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy_blast_out.csv:md5,e644132a7dafe4bc8fbb6f6271466f4e"
                    ]
                ],
                "csv": [
                    [
                        {
                            "id": "toy"
                        },
                        "toy.blastparse.csv:md5,0a87503881d0247b8a311f097a908dbb"
                    ]
                ],
                "png": [
                    [
                        {
                            "id": "toy"
                        },
                        [
                            "toy.alignment_plot.png:md5,3d15dd45ed55e7b34dfe2726fb867e77",
                            "toy.bitscore_plot.png:md5,6476ebee5eecfa61b8c004861b3eeff8",
                            "toy.contig_dot_plot.png:md5,88e63d8f54e05a909db2119a70673610",
                            "toy.hitlength_plot.png:md5,99d63c4ff172b695d7a1a5027a413d20"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2dc81e8053ed887ab8839a4dc176b569"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-06T11:24:47.002312611"
    }
}